pygmt/tests/test_grdlandmask.py

```diff
@@ -49,7 +49,7 @@ def test_grdlandmask_no_outgrid(expected_grid):
     """
     Test grdlandmask with no set outgrid.
     """
-    result = grdlandmask(spacing=1, region=[125, 130, 30, 35])
+    result = grdlandmask(spacing=1, region=[125, 130, 30, 35], cores=2)
```

Comment:
The Linux runner for GitHub Actions has 2 CPU cores according to https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources, but not sure if we should set cores to 2 or 1 for consistent benchmarking.

Reply:
Using two cores is faster and can help check the OpenMP feature in GMT, so I prefer to use two cores if possible.

BTW, the macOS runner for GitHub Actions has 3 cores, so instead of `cores=2`, `cores=True` will use all available cores.

Reply:
Is only 1 core being used if we don't set `cores`, or use `cores=None`? I'm reading https://docs.generic-mapping-tools.org/6.4/gmt.html#core-full which says `[default is to use all available cores]`, but is that only with `cores=True` (i.e. `-x` in GMT)?

The `benchmark.yml` workflow is running on Linux, but I guess we could set `cores=True` since this test would also run on macOS in `ci_test.yml`. My question was more around whether multi-threading with multiple-cores (2 or more) would provide a consistent benchmark compared to using only 1 core, but I guess we need to merge this PR to find out!

Reply:
> Is only 1 core being used if we don't set `cores`, or use `cores=None`? I'm reading https://docs.generic-mapping-tools.org/6.4/gmt.html#core-full which says `[default is to use all available cores]`, but is that only with `cores=True` (i.e. `-x` in GMT)?

Yes.

Reply:
> > Is only 1 core being used if we don't set `cores`, or use `cores=None`? I'm reading https://docs.generic-mapping-tools.org/6.4/gmt.html#core-full which says `[default is to use all available cores]`, but is that only with `cores=True` (i.e. `-x` in GMT)?
> 
> Yes.

Actually it uses all available cores if `-x` is not used.

```
$ gmt grdlandmask -R0/10/0/10 -Gabc.nc -I1/1 -Vi
grdlandmask [INFORMATION]: Enable all available threads (up to xxx)
```

Reply:
Just tried the GMT CLI. Looking at the `real` time, I expect it to be decreasing for the increasing number of cores, but apparently it doens't:
```
$ time gmt grdlandmask -R0/180/0/90 -I1m -Gabc.nc -x1

real	0m13.474s
user	0m16.252s
sys	0m6.025s

$ time gmt grdlandmask -R0/180/0/90 -I1m -Gabc.nc -x2

real	0m10.377s
user	0m22.465s
sys	0m5.969s

$ time gmt grdlandmask -R0/180/0/90 -I1m -Gabc.nc -x3

real	0m16.906s
user	0m51.020s
sys	0m6.215s

$ time gmt grdlandmask -R0/180/0/90 -I1m -Gabc.nc -x4

real	0m15.033s
user	0m59.322s
sys	0m6.043s

$ time gmt grdlandmask -R0/180/0/90 -I1m -Gabc.nc -x5

real	0m16.387s
user	1m19.829s
sys	0m6.144s

```

Reply:
Maybe set `cores=1` and see what the flame graph diff shows?

Reply:
> Maybe set `cores=1` and see what the flame graph diff shows?

The `test_grdlandmask_no_outgrid` benchmark speedup went from 3x to 9.6x :rofl: It might be that there is some overhead with using threading (2 cores) vs no threading (1 core). Here's the flame graph diff from https://codspeed.io/GenericMappingTools/pygmt/branches/limit-cores again:

![image](https://github.com/GenericMappingTools/pygmt/assets/23487320/20e0b3fc-dfd2-4bde-95e9-94185256b2c4)

The extra `gomp_team_barrier_wait_*` calls are still there, but also some other stuff (in blue).

Reply:
This reminds me of the warnings message I saw when running the tests in random order (see https://github.com/GenericMappingTools/pygmt/actions/runs/7396374248/job/20121388367?pr=2936):

```
pygmt/tests/test_session_management.py::test_session_multiprocessing
pygmt/tests/test_session_management.py::test_session_multiprocessing
  /home/runner/micromamba/envs/pygmt/lib/python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=2433) is multi-threaded, use of fork() may lead to deadlocks in the child.
    self.pid = os.fork()
```
There are some related discussions at https://discuss.python.org/t/concerns-regarding-deprecation-of-fork-with-alive-threads/33555 and related PR at https://github.com/python/cpython/issues/100228.

> os.fork() in a multi-threaded application is a likely source of deadlocks on many platforms. We should raise a warning when people call os.fork() from a process that we know has other threads running.

Not sure if it's related to the issue here.

Reply:
> ```
> pygmt/tests/test_session_management.py::test_session_multiprocessing
> pygmt/tests/test_session_management.py::test_session_multiprocessing
>   /home/runner/micromamba/envs/pygmt/lib/python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=2433) is multi-threaded, use of fork() may lead to deadlocks in the child.
>     self.pid = os.fork()
> ```

That [`test_session_multiprocessing`](https://github.com/GenericMappingTools/pygmt/blob/fbfed861018522225b0cae13404aa9428ba24395/pygmt/tests/test_session_management.py#L65-L88) function is only running `basemap` and `savefig` right? It shouldn't have any multi-threading going on, so not sure what's happening.

I tried running some benchmarks using cores from -1 (all cores) to 0 (no-multithreading) to 16 (all cores on my laptop). This was running `grdlandmask` with a spacing of 0.05 arc degrees. The results do vary between runs, but in general, more cores means less time:

![benchmark_grdlandmask_0.05arc_degree](https://github.com/GenericMappingTools/pygmt/assets/23487320/934a8136-0d30-4027-9b6d-8515c1119c87)

Code to reproduce:

```python
import time
import pandas as pd
import pygmt
import tqdm

timings = {}
for cores in tqdm.tqdm(range(-1, 16)):
    tic = time.perf_counter()
    pygmt.grdlandmask(spacing="0.05d", region=[0, 180, 0, 90], cores=cores)
    toc = time.perf_counter()
    timings[cores] = toc - tic

df = pd.DataFrame.from_dict(data=timings, orient="index", columns=["time"])

# %%
fig = pygmt.Figure()
region = pygmt.info(data=df.reset_index(), per_column=True, spacing=0.05)
fig.plot(
    x=df.index,
    y=df.time,
    region=region,
    pen="thick",
    frame=[
        "WSne+tBenchmark grdlandmask",
        "xaf+lCores",
        "yaf+lTime (s), lower is better",
    ],
)
fig.savefig("benchmark_grdlandmask.png")
fig.show()
```

If I change the spacing from 0.05 arc degrees to 1 arc degree, more cores actually take more time:

![benchmark_grdlandmask_1arc_degree](https://github.com/GenericMappingTools/pygmt/assets/23487320/d6b83f90-70d2-449f-baca-f1e0835f88f1)

For that `test_grdlandmask_no_outgrid` unit test which is using `spacing=1` (i.e. 1 arc degree), I think we should just use `cores=2`, since `cores=3` might not be much faster on such a low resolution. As long as we are benchmarking with a fixed number of core counts, it should be ok?

Reply:
sounds ok

Reply:
Ok, back to using `cores=2` in 97637d4dffa252ce3160cec452accfe5fec57fcc.